[FEAT/#106] 일기 작성하기 화면 텍스트 스캔하기 btn 카메라/갤러리로 연결되도록 기능 구현

Daljyeong opened the pull request:
## Related issue 🛠
- closed #106 

## Work Description ✏️
- '일기 작성하기' 화면의 텍스트 스캔하기 btn 기능 구현
- (+) 디자인 QA 반영 ('일기 작성하기' 화면의 TextField-사진 미리보기 사이 여백 수정)
- (+) 디자인 QA 반영 (ImageSelectBottomSheet 내 Icon 색상 변경)

## Screenshot 📸

https://github.com/user-attachments/assets/2d7dcb85-e75e-4929-be2f-baf5bf8f3bbe

디자인 QA 반영 전에 녹화해둔 영상이니 기능 위주로만 참고해 주세요..!!

## Uncompleted Tasks 😅
- [x] build.gradle.kts 버전 추출하기

## To Reviewers 📢
- 영상에서 TextField UI 이상한 부분은 민재옵이 수정해준 부분이랑 추후 합치면 해결됩니당
- ML Kit 기반으로 구현해서 영어 텍스트만 인식 가능합니다
- 애뮬레이터랑 30분 동안 싸우고 왔어요... 웬만하면 실기기로 테스트해보는 걸 추천합니다 ㅠ

Daljyeong committed:
[FEAT/#106] DiaryWriteScreen 텍스트 스캔 기능 구현

Daljyeong committed:
[FIX/#106] DiaryWriteScreen 디자인 QA 반영 (TextField-사진 미리보기 사이 여백 수정)

Daljyeong committed:
[DEL/#106] AndroidManifest.xml 인터넷 권한 중복 선언한 부분 삭제

Daljyeong committed:
Merge remote-tracking branch 'origin/develop' into feat/#106-diarywrite-text-scan

Hyobeen-Park commented on presentation/diarywrite/build.gradle.kts:
```diff
@@ -11,4 +11,8 @@ android {
+
+    // ML Kit
+    implementation("com.google.mlkit:text-recognition:16.0.1")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-play-services:1.7.1")
```
P1: 버전카탈로그 사용해주세요~

Hyobeen-Park commented on app/src/main/res/xml/file_paths.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<paths xmlns:android="http://schemas.android.com/apk/res/android">
+    <cache-path
+        name="camera_images"
+        path="." />
```
오홍 이런거 해줘야 하는구낭!!

Hyobeen-Park commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteScreen.kt:
```diff
@@ -113,6 +157,16 @@ internal fun DiaryWriteRoute(
 }
 
+private fun createImageUri(context: Context): Uri {
+    val imageFile = File.createTempFile("camera_", ".jpg", context.cacheDir)
```
P3: 잘 모르긴 하는데 임시 파일 만드는거 맞나여?? 이거 알아서 다시 메모리 해제되낭??!?

Daljyeong replied:
추후 TODO로 넘기는 식으로 반영 완..!

Hyobeen-Park commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteViewModel.kt:
```diff
@@ -58,4 +65,29 @@ internal class DiaryWriteViewModel @Inject constructor(
+                    it.copy(diaryText = filteredText)
+                }
+            } catch (e: Exception) {
+                Log.e("DiaryWriteViewModel", "Text recognition failed", e)
```
P1: 꼭 필요하다면 로그 말고 팀버쓰기~~ 혹시 차이점 기억 안나면 저 다시 찾아오세요ㅋㅋ

Hyobeen-Park commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteViewModel.kt:
```diff
@@ -15,14 +20,16 @@ import kotlinx.coroutines.flow.StateFlow
     savedStateHandle: SavedStateHandle,
     private val calendarRepository: CalendarRepository
-) : ViewModel() {
+) : AndroidViewModel(application) {
```
P1: 우엥 이거 `ViewModel()`로는 안되는건가요?????

Daljyeong replied:
헿 리팩 조금 미뤄뒀는데 텍스트 스캔 부분을 외부로 빼서 주입받으면 ViewModel()로 처리가 가능하네용 ㅎㅅㅎ
리팩해서 반영해두겠습니다 !

nhyeonii commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteScreen.kt:
```diff
@@ -63,10 +70,41 @@ internal fun DiaryWriteRoute(
+        contract = ActivityResultContracts.PickVisualMedia()
+    ) { uri ->
+        uri?.let {
+            viewModel.extractTextFromImage(it)
```
p2: 요것두 viewModel::extractTextFromImage이렇게 작성할 수 있지 않나요 ?!?

Daljyeong replied:
`uri?.let(viewModel::extractTextFromImage)`로 반영 완료 !

Hyobeen-Park commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteViewModel.kt:
```diff
@@ -58,4 +65,29 @@ internal class DiaryWriteViewModel @Inject constructor(
+    fun extractTextFromImage(uri: Uri) {
+        viewModelScope.launch {
+            try {
+                val context = getApplication<Application>().applicationContext
```
P1: 흠.. 이게 조금 마음에 걸리는데 context 받아올 수 있는 다른 방법 찾아보면 좋을 것 같네요!!

Hyobeen-Park reviewed (commented):
우와 신기해!!

Daljyeong committed:
[FIX/#106] DiaryWriteScreen 디자인 QA 반영 (ImageSelectBottomSheet 내 Icon 색상 변경)

Daljyeong committed:
Merge remote-tracking branch 'origin/develop' into feat/#106-diarywrite-text-scan

Daljyeong committed:
Merge remote-tracking branch 'origin/develop' into feat/#106-diarywrite-text-scan

Daljyeong committed:
[MOD/#106] nahyeonii 코드 리뷰 반영

Daljyeong committed:
[MOD/#106] Hyobeen-Park 코드 리뷰 반영

nhyeonii reviewed (approved):
확인 완료 !!! 고생햇어요 신기하다잉

Daljyeong committed:
[MOD/#106] TextRecognitionManager.kt 생성해 텍스트 스캔 로직 분리

Daljyeong committed:
Merge remote-tracking branch 'origin/develop' into feat/#106-diarywrite-text-scan

# Conflicts:
#	presentation/diarywrite/build.gradle.kts
#	presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteScreen.kt
#	presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteViewModel.kt

Daljyeong committed:
Merge remote-tracking branch 'origin/develop' into feat/#106-diarywrite-text-scan

nahy-512 commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteScreen.kt:
```diff
@@ -139,6 +191,8 @@ private fun DiaryWriteScreen(
     onDiaryTextChanged: (String) -> Unit,
     diaryImageUri: Uri?,
     onDiaryImageUriChanged: (Uri?) -> Unit,
+    onBottomSheetCameraClicked: () -> Unit,
+    onBottomSheetGalleryClicked: () -> Unit,
```
p3
아래는 Click으로 끝나는데 여기 둘은 Clicked이네용..!
둘이 다른 이유가 있을까요?

nahy-512 commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteViewModel.kt:
```diff
@@ -75,4 +78,22 @@ internal class DiaryWriteViewModel @Inject constructor(
+                    extractedText.substring(0, 1000)
+                } else {
+                    extractedText
+                }
```
p3
```suggestion
                val filteredText = extractedText.run {
        if (length > 1000) this.take(1000) else this
    }
```
이것처럼 `substring` 말고도 `take`를 사용하는 방법도 있어요~!

nahy-512 reviewed (approved):
텍스트 스캔 넘 신기하다ㅎ.ㅎ
고생하셨습니다~!!!!

angryPodo committed:
[MOD/#106] 전체적으로 리펙토링
- TextRecognitionManager제거 및 ViewModel 통함
- Dispatchers.IO로 스레드 분리
- 사진 촬영 후 임시 파일 삭제 로직 추가

Hyobeen-Park commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteScreen.kt:
```diff
@@ -63,14 +70,47 @@ internal fun DiaryWriteRoute(
+        contract = ActivityResultContracts.RequestPermission()
+    ) { isGranted ->
+        if (isGranted) {
+            val (uri, file) = createTempImageFile(context)
```
👍🏻

angryPodo reviewed (approved):
구현으로 리뷰 대체!!! 잘햇어ㅎㅎㅎ 🚀🚀🚀🚀🚀

Hyobeen-Park commented on presentation/diarywrite/src/main/java/com/hilingual/presentation/diarywrite/DiaryWriteViewModel.kt:
```diff
@@ -75,4 +86,36 @@ internal class DiaryWriteViewModel @Inject constructor(
+                }
+            } finally {
+                withContext(Dispatchers.IO) {
+                    if (tempFileToDelete?.exists() == true) {
+                        tempFileToDelete.delete()
+                    }
```
개굿

Hyobeen-Park reviewed (approved):
어푸푸🚀🚀
